pkg/registry/core/service/storage/storage.go

```diff
@@ -248,64 +247,44 @@ func (r *REST) defaultOnReadService(service *api.Service) {
-
-		// headless+selectorless
-		// headless+selectorless takes both families. Why?
-		// at this stage we don't know what kind of endpoints (specifically their IPFamilies) the
-		// user has assigned to this selectorless service. We assume it has dualstack and we default
-		// it to PreferDualStack on any cluster (single or dualstack configured).
-		if len(service.Spec.Selector) == 0 {
-			service.Spec.IPFamilyPolicy = &preferDualStack
-			if r.primaryIPFamily == api.IPv4Protocol {
-				service.Spec.IPFamilies = append(service.Spec.IPFamilies, api.IPv6Protocol)
-			} else {
-				service.Spec.IPFamilies = append(service.Spec.IPFamilies, api.IPv4Protocol)
-			}
-		} else {
-			// headless w/ selector
-			// this service type follows cluster configuration. this service (selector based) uses a
-			// selector and will have to follow how the cluster is configured. If the cluster is
-			// configured to dual stack then the service defaults to PreferDualStack. Otherwise we
-			// default it to SingleStack.
-			if r.secondaryIPFamily != "" {
-				service.Spec.IPFamilies = append(service.Spec.IPFamilies, r.secondaryIPFamily)
-				service.Spec.IPFamilyPolicy = &preferDualStack
+			if len(service.Spec.Selector) == 0 {
+				// Headless + selectorless is a special-case.
+				//
+				// At this stage we don't know what kind of endpoints (specifically
+				// their IPFamilies) the user has assigned to this selectorless
+				// service. We assume it has dual-stack and we default it to
+				// RequireDualStack on any cluster (single- or dual-stack
+				// configured).
+				service.Spec.IPFamilyPolicy = &requireDualStack
+				service.Spec.IPFamilies = []api.IPFamily{r.primaryIPFamily, otherFamily(r.primaryIPFamily)}
 			} else {
+				// Headless + selector - default to single.
 				service.Spec.IPFamilyPolicy = &singleStack
+				service.Spec.IPFamilies = []api.IPFamily{r.primaryIPFamily}
 			}
-		}
-	} else {
-		// headful
-		// make sure a slice exists to receive the families
-		service.Spec.IPFamilies = make([]api.IPFamily, len(service.Spec.ClusterIPs), len(service.Spec.ClusterIPs))
-		for idx, ip := range service.Spec.ClusterIPs {
-			if netutil.IsIPv6String(ip) {
-				service.Spec.IPFamilies[idx] = api.IPv6Protocol
-			} else {
-				service.Spec.IPFamilies[idx] = api.IPv4Protocol
-			}
-
-			if len(service.Spec.IPFamilies) == 1 {
-				service.Spec.IPFamilyPolicy = &singleStack
-			} else if len(service.Spec.IPFamilies) == 2 {
-				service.Spec.IPFamilyPolicy = &requireDualStack
+		} else {
+			// Headful: init ipFamilies from clusterIPs
+			service.Spec.IPFamilies = make([]api.IPFamily, len(service.Spec.ClusterIPs))
+			for idx, ip := range service.Spec.ClusterIPs {
+				if netutil.IsIPv6String(ip) {
+					service.Spec.IPFamilies[idx] = api.IPv6Protocol
+				} else {
+					service.Spec.IPFamilies[idx] = api.IPv4Protocol
+				}
+				if len(service.Spec.IPFamilies) == 1 {
+					service.Spec.IPFamilyPolicy = &singleStack
```

Comment:
what if this was already set?

Reply:
I don't know how IPFamilyPolicy would  be set but ipFamilies would not?